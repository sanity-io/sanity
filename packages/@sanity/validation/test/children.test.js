const {Rule} = require('../src')

describe('child rules', () => {
  // --- ALL ---
  test('all() rules - single failure', async () => {
    const rule = Rule.string().all([
      Rule.string()
        .regex(/^[A-Z]/)
        .error('Must start with an uppercase character'),
      Rule.string()
        .regex(/[a-z]+/)
        .error('Must follow with lowercase characters'),
    ])

    await expect(rule.validate('Sanity')).resolves.toMatchSnapshot('all() rules - match')
    await expect(rule.validate('moop')).resolves.toMatchSnapshot(
      'all() rules - single failure, custom message'
    )
  })

  test('all() rules - multiple failures', async () => {
    const rule = Rule.string().all([
      Rule.string()
        .regex(/^[A-Z]/)
        .error('Must start with an uppercase character'),
      Rule.string().min(5).error('5 chars or more'),
      Rule.string()
        .regex(/[a-z]+/)
        .error('Must follow with lowercase characters'),
    ])

    await expect(rule.validate('Sanity')).resolves.toMatchSnapshot('all() rules - match')
    await expect(rule.validate('moop')).resolves.toMatchSnapshot(
      'all() rules - multiple failures, custom messages'
    )
  })

  test('all() rules - single failure, custom, common error', async () => {
    const rule = Rule.string()
      .all([Rule.string().regex(/^[A-Z]/), Rule.string().regex(/[a-z]+/)])
      .error('Needs to start with a capital letter and then follow with lowercase characters')

    await expect(rule.validate('moop')).resolves.toMatchSnapshot(
      'all() rules - single failure, common error'
    )
  })

  test('all() rules - single failure, custom, common error', async () => {
    const rule = Rule.string()
      .all([Rule.string().regex(/^[A-Z]/), Rule.string().min(5), Rule.string().regex(/[a-z]+/)])
      .error('Needs to be a capital letter followed by at least 4 lowercase characters')

    await expect(rule.validate('moop')).resolves.toMatchSnapshot(
      'all() rules - multiple failures, common error'
    )
  })

  // --- EITHER ---
  test('either() rules - single failure', async () => {
    const rule = Rule.string().either([
      Rule.string()
        .regex(/^rgb(\d+,\s*\d+,\s*\d+)$/)
        .error('Must be rgb(num, num, num) format'),
      Rule.string()
        .regex(/^#([a-f0-9]{3}|[a-f0-9]{6})$/)
        .error('Must be hex color with #-prefix'),
    ])

    await expect(rule.validate('rgb(16, 22, 133)')).resolves.toMatchSnapshot(
      'either() rules - match'
    )
    await expect(rule.validate('#bf')).resolves.toMatchSnapshot(
      'either() rules - single failure, custom message'
    )
  })

  test('either() rules - all matches', async () => {
    const rule = Rule.string().either([
      Rule.string().regex(/^R/).error('Must start with a capital R'),
      Rule.string().regex(/ed$/).error('Must end with "ed"'),
    ])

    await expect(rule.validate('Red')).resolves.toMatchSnapshot('either() rules - all match')
    await expect(rule.validate('nope')).resolves.toMatchSnapshot('either() rules - no matches')
  })

  test('either() rules - all fail, custom, common error', async () => {
    const rule = Rule.string()
      .either([Rule.string().regex(/^[A-Z]/), Rule.string().regex(/^i[A-Z]/)])
      .error("Needs to start with a capital letter, unless it's an iProduct")

    await expect(rule.validate('mopatis!')).resolves.toMatchSnapshot(
      'either() rules - all fail, common error'
    )
  })
})
