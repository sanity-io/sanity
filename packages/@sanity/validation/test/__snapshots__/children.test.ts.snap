// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`child rules all() rules - multiple failures: all() rules - match 1`] = `Array []`;

exports[`child rules all() rules - multiple failures: all() rules - multiple failures, custom messages 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": Array [
        Object {
          "item": ValidationError {
            "children": undefined,
            "message": "Must start with an uppercase character",
            "operation": undefined,
            "paths": Array [],
          },
          "level": "error",
          "path": Array [],
          "type": "validation",
        },
        Object {
          "item": ValidationError {
            "children": undefined,
            "message": "5 chars or more",
            "operation": undefined,
            "paths": Array [],
          },
          "level": "error",
          "path": Array [],
          "type": "validation",
        },
      ],
      "message": "[Must start with an uppercase character - AND - 5 chars or more]",
      "operation": "AND",
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`child rules all() rules - single failure, custom, common error: all() rules - multiple failures, common error 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": Array [
        Object {
          "item": ValidationError {
            "children": undefined,
            "message": "Does not match \\"/^[A-Z]/\\"-pattern",
            "operation": undefined,
            "paths": Array [],
          },
          "level": "error",
          "path": Array [],
          "type": "validation",
        },
        Object {
          "item": ValidationError {
            "children": undefined,
            "message": "Must be at least 5 characters long",
            "operation": undefined,
            "paths": Array [],
          },
          "level": "error",
          "path": Array [],
          "type": "validation",
        },
      ],
      "message": "Needs to be a capital letter followed by at least 4 lowercase characters",
      "operation": "AND",
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`child rules all() rules - single failure, custom, common error: all() rules - single failure, common error 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Needs to start with a capital letter and then follow with lowercase characters",
      "operation": "AND",
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`child rules all() rules - single failure: all() rules - match 1`] = `Array []`;

exports[`child rules all() rules - single failure: all() rules - single failure, custom message 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must start with an uppercase character",
      "operation": "AND",
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`child rules either() rules - all fail, custom, common error: either() rules - all fail, common error 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": Array [
        Object {
          "item": ValidationError {
            "children": undefined,
            "message": "Does not match \\"/^[A-Z]/\\"-pattern",
            "operation": undefined,
            "paths": Array [],
          },
          "level": "error",
          "path": Array [],
          "type": "validation",
        },
        Object {
          "item": ValidationError {
            "children": undefined,
            "message": "Does not match \\"/^i[A-Z]/\\"-pattern",
            "operation": undefined,
            "paths": Array [],
          },
          "level": "error",
          "path": Array [],
          "type": "validation",
        },
      ],
      "message": "Needs to start with a capital letter, unless it's an iProduct",
      "operation": "OR",
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`child rules either() rules - all matches: either() rules - all match 1`] = `Array []`;

exports[`child rules either() rules - all matches: either() rules - no matches 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": Array [
        Object {
          "item": ValidationError {
            "children": undefined,
            "message": "Must start with a capital R",
            "operation": undefined,
            "paths": Array [],
          },
          "level": "error",
          "path": Array [],
          "type": "validation",
        },
        Object {
          "item": ValidationError {
            "children": undefined,
            "message": "Must end with \\"ed\\"",
            "operation": undefined,
            "paths": Array [],
          },
          "level": "error",
          "path": Array [],
          "type": "validation",
        },
      ],
      "message": "[Must start with a capital R - OR - Must end with \\"ed\\"]",
      "operation": "OR",
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`child rules either() rules - single failure: either() rules - match 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": Array [
        Object {
          "item": ValidationError {
            "children": undefined,
            "message": "Must be rgb(num, num, num) format",
            "operation": undefined,
            "paths": Array [],
          },
          "level": "error",
          "path": Array [],
          "type": "validation",
        },
        Object {
          "item": ValidationError {
            "children": undefined,
            "message": "Must be hex color with #-prefix",
            "operation": undefined,
            "paths": Array [],
          },
          "level": "error",
          "path": Array [],
          "type": "validation",
        },
      ],
      "message": "[Must be rgb(num, num, num) format - OR - Must be hex color with #-prefix]",
      "operation": "OR",
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`child rules either() rules - single failure: either() rules - single failure, custom message 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": Array [
        Object {
          "item": ValidationError {
            "children": undefined,
            "message": "Must be rgb(num, num, num) format",
            "operation": undefined,
            "paths": Array [],
          },
          "level": "error",
          "path": Array [],
          "type": "validation",
        },
        Object {
          "item": ValidationError {
            "children": undefined,
            "message": "Must be hex color with #-prefix",
            "operation": undefined,
            "paths": Array [],
          },
          "level": "error",
          "path": Array [],
          "type": "validation",
        },
      ],
      "message": "[Must be rgb(num, num, num) format - OR - Must be hex color with #-prefix]",
      "operation": "OR",
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;
